

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>VariantExtractor &mdash; variant-extractor 4.0.10 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=b8a5ba81"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="variant_extractor API" href="variant_extractor.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            variant-extractor
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="variant_extractor.html">variant_extractor API</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">CHANGELOG</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">variant-extractor</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">VariantExtractor<!-- omit in toc --></li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="variantextractor">
<h1>VariantExtractor<!-- omit in toc --><a class="headerlink" href="#variantextractor" title="Link to this heading"></a></h1>
<p><a class="reference external" href="https://zenodo.org/doi/10.5281/zenodo.12755170"><img alt="DOI" src="https://zenodo.org/badge/476634824.svg" /></a></p>
<p>R. Martín et al., “ONCOLINER: A new solution for monitoring, improving, and harmonizing somatic variant calling across genomic oncology centers,” <em>Cell Genomics</em>, vol. 4, no. 9. Elsevier BV, p. 100639, Sep. 2024. <a class="reference external" href="https://doi.org/10.1016/j.xgen.2024.100639">doi: 10.1016/j.xgen.2024.100639</a></p>
<p><strong>Deterministic and standard extractor of indels, SNVs and structural variants (SVs)</strong> from VCF files built under the frame of <a class="reference external" href="https://eucancan.com/">EUCANCan</a>’s second work package. VariantExtractor is a Python package (<strong>requires Python version 3.6 or higher</strong>) and provides a set of data structures and functions to extract variants from VCF files in a <strong>deterministic and standard</strong> way while <a class="reference internal" href="#variantrecord"><span class="xref myst">adding information</span></a> to facilitate afterwards processing. It homogenizes <a class="reference internal" href="#multiallelic-variants"><span class="xref myst">multiallelic variants</span></a>, <a class="reference internal" href="#snvs"><span class="xref myst">MNPs</span></a> and <a class="reference internal" href="#structural-variants"><span class="xref myst">SVs</span></a> (including <a class="reference internal" href="#imprecise-paired-breakends"><span class="xref myst">imprecise paired breakends</span></a> and <a class="reference internal" href="#single-breakends"><span class="xref myst">single breakends</span></a>). The package is designed to be used in a pipeline, where the variants are ingested from VCF files and then used in downstream analysis. Check the <a class="reference external" href="https://eucancan.github.io/variant-extractor/">available documentation</a> for more information.</p>
<p>While there is somewhat of an agreement on how to label the SNVs and indels variants, this is not the case for the structural variants. In the current scenario, different labeling between variant callers makes comparisons between structural variants difficult. This package provides an unified interface to extract variants (included structural variants) from VCFs generated by different variant callers. Apart from reading the VCF file, VariantExtractor <strong>adds a preprocessing layer to homogenize the variants</strong> extracted from the file. This way, the variants can be used in downstream analysis in a consistent way. For more information about the homogenization process, check the <a class="reference internal" href="#homogenization-rules"><span class="xref myst">homogenization rules</span></a> section.</p>
<section id="table-of-contents">
<h2>Table of contents<!-- omit in toc --><a class="headerlink" href="#table-of-contents" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="#getting-started"><span class="xref myst">Getting started</span></a></p>
<ul>
<li><p><a class="reference internal" href="#installation"><span class="xref myst">Installation</span></a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#usage"><span class="xref myst">Usage</span></a></p></li>
<li><p><a class="reference internal" href="#variantrecord"><span class="xref myst">VariantRecord</span></a></p>
<ul>
<li><p><a class="reference internal" href="#varianttype"><span class="xref myst">VariantType</span></a></p></li>
<li><p><a class="reference internal" href="#breakendsvrecord"><span class="xref myst">BreakendSVRecord</span></a></p></li>
<li><p><a class="reference internal" href="#shorthandsvrecord"><span class="xref myst">ShorthandSVRecord</span></a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#homogenization-rules"><span class="xref myst">Homogenization rules</span></a></p>
<ul>
<li><p><a class="reference internal" href="#multiallelic-variants"><span class="xref myst">Multiallelic variants</span></a></p></li>
<li><p><a class="reference internal" href="#snvs"><span class="xref myst">SNVs</span></a></p></li>
<li><p><a class="reference internal" href="#structural-variants"><span class="xref myst">Structural variants</span></a></p>
<ul>
<li><p><a class="reference internal" href="#breakend-vs-shorthand-notation"><span class="xref myst">Breakend vs shorthand notation</span></a></p></li>
<li><p><a class="reference internal" href="#paired-breakends"><span class="xref myst">Paired breakends</span></a></p></li>
<li><p><a class="reference internal" href="#inferred-breakend-pairs"><span class="xref myst">Inferred breakend pairs</span></a></p></li>
<li><p><a class="reference internal" href="#imprecise-paired-breakends"><span class="xref myst">Imprecise paired breakends</span></a></p></li>
<li><p><a class="reference internal" href="#single-breakends"><span class="xref myst">Single breakends</span></a></p></li>
</ul>
</li>
</ul>
</li>
<li><p><a class="reference internal" href="#dependencies"><span class="xref myst">Dependencies</span></a></p></li>
</ul>
</section>
<section id="getting-started">
<h2>Getting started<a class="headerlink" href="#getting-started" title="Link to this heading"></a></h2>
<section id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Link to this heading"></a></h3>
<p>VariantExtractor is available on PyPI and can be installed using <code class="docutils literal notranslate"><span class="pre">pip</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>variant-extractor
</pre></div>
</div>
</section>
</section>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Link to this heading"></a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the package</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">variant_extractor</span><span class="w"> </span><span class="kn">import</span> <span class="n">VariantExtractor</span>

<span class="c1"># Create a new instance of the class</span>
<span class="n">extractor</span> <span class="o">=</span> <span class="n">VariantExtractor</span><span class="p">(</span><span class="s1">&#39;/path/to/file.vcf&#39;</span><span class="p">)</span>
<span class="c1"># Iterate through the variants</span>
<span class="k">for</span> <span class="n">variant_record</span> <span class="ow">in</span> <span class="n">extractor</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Found variant of type </span><span class="si">{</span><span class="n">variant_record</span><span class="o">.</span><span class="n">variant_type</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="n">variant_record</span><span class="o">.</span><span class="n">contig</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">variant_record</span><span class="o">.</span><span class="n">pos</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import the package</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">variant_extractor</span><span class="w"> </span><span class="kn">import</span> <span class="n">VariantExtractor</span>

<span class="c1"># Create a new instance of the class</span>
<span class="n">extractor</span> <span class="o">=</span> <span class="n">VariantExtractor</span><span class="p">(</span><span class="s1">&#39;/path/to/file.vcf&#39;</span><span class="p">)</span>

<span class="c1"># Save variants to a CSV file</span>
<span class="n">extractor</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;variant_record_obj&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;/path/to/output.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>For a more complete list of examples, check the <a class="reference internal" href="#./examples/"><span class="xref myst">examples</span></a> directory. This folder also includes an example of a <a class="reference internal" href="#examples/normalize_vcf.py"><span class="xref myst">script for normalizing VCF files</span></a> following the <a class="reference internal" href="#homogenization-rules"><span class="xref myst">homogenization rules</span></a>.</p>
</section>
<section id="variantrecord">
<h2>VariantRecord<a class="headerlink" href="#variantrecord" title="Link to this heading"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">VariantExtractor</span></code> constructor returns a generator of <code class="docutils literal notranslate"><span class="pre">VariantRecord</span></code> instances. The <code class="docutils literal notranslate"><span class="pre">VariantRecord</span></code> class is a container for the information contained in a VCF record plus some extra useful information.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">contig</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">str</span></code></p></td>
<td><p>Contig name</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pos</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">int</span></code></p></td>
<td><p>Position on the contig</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">end</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">int</span></code></p></td>
<td><p>End position of the variant in the contig (same as <code class="docutils literal notranslate"><span class="pre">pos</span></code> for TRA and SNV)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">length</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">int</span></code></p></td>
<td><p>Length of the variant</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">id</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">Optional[str]</span></code></p></td>
<td><p>Record identifier</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">ref</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">str</span></code></p></td>
<td><p>Reference sequence</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">alt</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">str</span></code></p></td>
<td><p>Alternative sequence</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">qual</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">Optional[float]</span></code></p></td>
<td><p>Quality score for the assertion made in ALT</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">filter</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">List[str]</span></code></p></td>
<td><p>Filter status. <code class="docutils literal notranslate"><span class="pre">PASS</span></code> if this position has passed all filters. Otherwise, it contains the filters that failed</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">info</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">Dict[str,</span> <span class="pre">Any]</span></code></p></td>
<td><p>Additional information</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">format</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">List[str]</span></code></p></td>
<td><p>Specifies data types and order of the genotype information</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">samples</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">Dict[str,</span> <span class="pre">Dict[str,</span> <span class="pre">Any]]</span></code></p></td>
<td><p>Genotype information for each sample</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">variant_type</span></code></p></td>
<td><p><a class="reference internal" href="#varianttype"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></span></a></p></td>
<td><p>Variant type inferred</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">alt_sv_breakend</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">Optional[</span></code><a class="reference internal" href="#brekendsvrecord"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">BreakendSVRecord</span></code></span></a><code class="docutils literal notranslate"><span class="pre">]</span></code></p></td>
<td><p>Breakend SV info, present only for SVs with breakend notation. For example, <code class="docutils literal notranslate"><span class="pre">G]17:198982]</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">alt_sv_shorthand</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">Optional[</span></code><a class="reference internal" href="#shorthandsvrecord"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">ShorthandSVRecord</span></code></span></a><code class="docutils literal notranslate"><span class="pre">]</span></code></p></td>
<td><p>Shorthand SV info, present only for SVs with shorthand notation. For example, <code class="docutils literal notranslate"><span class="pre">&lt;DUP:TANDEM&gt;</span></code></p></td>
</tr>
</tbody>
</table>
<section id="varianttype">
<h3>VariantType<a class="headerlink" href="#varianttype" title="Link to this heading"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">VariantType</span></code> enum describes the type of the variant. For structural variants, it is inferred <strong>only</strong> from the breakend notation (or shorthand notation). It does not take into account any <code class="docutils literal notranslate"><span class="pre">INFO</span></code> field (<code class="docutils literal notranslate"><span class="pre">SVTYPE</span></code> nor <code class="docutils literal notranslate"><span class="pre">EVENTYPE</span></code>) that might be added by the variant caller afterwards.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>Variant name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>A</p></td>
<td><p>G</p></td>
<td><p>SNV</p></td>
<td><p>Single nucleotide variant</p></td>
</tr>
<tr class="row-odd"><td><p>AGTG</p></td>
<td><p>A</p></td>
<td><p>DEL</p></td>
<td><p>Deletion</p></td>
</tr>
<tr class="row-even"><td><p>A</p></td>
<td><p>A[1:20[ or &lt;DEL&gt;</p></td>
<td><p>DEL</p></td>
<td><p>Deletion</p></td>
</tr>
<tr class="row-odd"><td><p>A</p></td>
<td><p>ACCT or &lt;INS&gt;</p></td>
<td><p>INS</p></td>
<td><p>Insertion</p></td>
</tr>
<tr class="row-even"><td><p>A</p></td>
<td><p>]1:20]A or &lt;DUP&gt;</p></td>
<td><p>DUP</p></td>
<td><p>Duplication</p></td>
</tr>
<tr class="row-odd"><td><p>A</p></td>
<td><p>A]1:20] or [1:20[A</p></td>
<td><p>INV</p></td>
<td><p>Inversion. <strong><a class="reference internal" href="#the-special-case-of-inv"><span class="xref myst">&lt;INV&gt; is a special case</span></a></strong></p></td>
</tr>
<tr class="row-even"><td><p>A</p></td>
<td><p>&lt;CNV&gt;</p></td>
<td><p>CNV</p></td>
<td><p>Copy number variation</p></td>
</tr>
<tr class="row-odd"><td><p>A</p></td>
<td><p>A]X:20] or A[X:20[ or ]X:20]A or [X:20[A</p></td>
<td><p>TRA</p></td>
<td><p>Translocation</p></td>
</tr>
<tr class="row-even"><td><p>A</p></td>
<td><p>A. or .A</p></td>
<td><p>SGL</p></td>
<td><p>Single breakend</p></td>
</tr>
</tbody>
</table>
</section>
<section id="breakendsvrecord">
<h3>BreakendSVRecord<a class="headerlink" href="#breakendsvrecord" title="Link to this heading"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">BreakendSVRecord</span></code> class is a container for the information contained in a VCF record for SVs with breakend notation.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">prefix</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">Optional[str]</span></code></p></td>
<td><p>Prefix of the SV record with breakend notation. For example, for <code class="docutils literal notranslate"><span class="pre">G]17:198982]</span></code> the prefix will be <code class="docutils literal notranslate"><span class="pre">G</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">bracket</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">str</span></code></p></td>
<td><p>Bracket of the SV record with breakend notation. For example, for <code class="docutils literal notranslate"><span class="pre">G]17:198982]</span></code> the bracket will be <code class="docutils literal notranslate"><span class="pre">]</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">contig</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">str</span></code></p></td>
<td><p>Contig of the SV record with breakend notation. For example, for <code class="docutils literal notranslate"><span class="pre">G]17:198982]</span></code> the contig will be <code class="docutils literal notranslate"><span class="pre">17</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pos</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">int</span></code></p></td>
<td><p>Position of the SV record with breakend notation. For example, for <code class="docutils literal notranslate"><span class="pre">G]17:198982]</span></code> the position will be <code class="docutils literal notranslate"><span class="pre">198982</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">suffix</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">Optional[str]</span></code></p></td>
<td><p>Suffix of the SV record with breakend notation. For example, for <code class="docutils literal notranslate"><span class="pre">G]17:198982]</span></code> the suffix will be <code class="docutils literal notranslate"><span class="pre">None</span></code></p></td>
</tr>
</tbody>
</table>
</section>
<section id="shorthandsvrecord">
<h3>ShorthandSVRecord<a class="headerlink" href="#shorthandsvrecord" title="Link to this heading"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">ShorthandSVRecord</span></code> class is a container for the information contained in a VCF record for SVs with shorthand notation.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Property</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">type</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">str</span></code></p></td>
<td><p>Type of the SV record with shorthand notation. One of the following, <code class="docutils literal notranslate"><span class="pre">'DEL'</span></code>, <code class="docutils literal notranslate"><span class="pre">'INS'</span></code>, <code class="docutils literal notranslate"><span class="pre">'DUP'</span></code>, <code class="docutils literal notranslate"><span class="pre">'INV'</span></code> or <code class="docutils literal notranslate"><span class="pre">'CNV'</span></code>. For example, for <code class="docutils literal notranslate"><span class="pre">&lt;DUP:TANDEM&gt;</span></code> the type will be <code class="docutils literal notranslate"><span class="pre">DUP</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">extra</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">List[str]</span></code></p></td>
<td><p>Extra information of the SV. For example, for <code class="docutils literal notranslate"><span class="pre">&lt;DUP:TANDEM:AA&gt;</span></code> the extra will be <code class="docutils literal notranslate"><span class="pre">['TANDEM',</span> <span class="pre">'AA']</span></code></p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="homogenization-rules">
<h2>Homogenization rules<a class="headerlink" href="#homogenization-rules" title="Link to this heading"></a></h2>
<p>VariantExtractor provides a unified interface to extract variants (included structural variants) from VCF files generated by different variant callers. The variants are homogenized and returned applying the following rules:</p>
<section id="multiallelic-variants">
<h3>Multiallelic variants<a class="headerlink" href="#multiallelic-variants" title="Link to this heading"></a></h3>
<p>An entry with multiple <code class="docutils literal notranslate"><span class="pre">ALT</span></code> sequences (multiallelic) is divided into multiple entries with a single <code class="docutils literal notranslate"><span class="pre">ALT</span></code> field. This entries with a single <code class="docutils literal notranslate"><span class="pre">ALT</span></code> field are then processed with the rest of the homogeneization rules. For example:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>1</p></td>
<td><p>multiallelic_1</p></td>
<td><p>G</p></td>
<td><p>C,T</p></td>
<td><p>PASS</p></td>
</tr>
</tbody>
</table>
<p>is returned as:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p><a class="reference internal" href="#varianttype"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></span></a></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>1</p></td>
<td><p>multiallelic_1_0</p></td>
<td><p>G</p></td>
<td><p>C</p></td>
<td><p>PASS</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>1</p></td>
<td><p>multiallelic_1_1</p></td>
<td><p>G</p></td>
<td><p>T</p></td>
<td><p>PASS</p></td>
<td><p>SNV</p></td>
</tr>
</tbody>
</table>
</section>
<section id="snvs">
<h3>SNVs<a class="headerlink" href="#snvs" title="Link to this heading"></a></h3>
<p>Entries with <code class="docutils literal notranslate"><span class="pre">REF/ALT</span></code> of the same lenghts are treated like SNVs. If the <code class="docutils literal notranslate"><span class="pre">REF/ALT</span></code> sequences are more than one nucleotide (MNPs), they are divided into multiple atomic SNVs. For example:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>1</p></td>
<td><p>snv_1</p></td>
<td><p>C</p></td>
<td><p>G</p></td>
<td><p>PASS</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>3</p></td>
<td><p>mnp_1</p></td>
<td><p>TAG</p></td>
<td><p>AGT</p></td>
<td><p>PASS</p></td>
</tr>
</tbody>
</table>
<p>are returned as:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p><a class="reference internal" href="#varianttype"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></span></a></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>3</p></td>
<td><p>snv_1</p></td>
<td><p>C</p></td>
<td><p>G</p></td>
<td><p>PASS</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>3</p></td>
<td><p>mnp_1_0</p></td>
<td><p>T</p></td>
<td><p>A</p></td>
<td><p>PASS</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>4</p></td>
<td><p>mnp_1_1</p></td>
<td><p>A</p></td>
<td><p>G</p></td>
<td><p>PASS</p></td>
<td><p>SNV</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>5</p></td>
<td><p>mnp_1_2</p></td>
<td><p>G</p></td>
<td><p>T</p></td>
<td><p>PASS</p></td>
<td><p>SNV</p></td>
</tr>
</tbody>
</table>
<!-- ### Compound indels
All entries with the `REF/ALT` of different lengths are treated as compound indels (or complex indels). They are left-trimmed and divided into multiple atomic SNVs and an insertion (INS) or a deletion (DEL). If the `REF` sequence is longer than the `ALT` sequence, it is considered a deletion. If the `REF` sequence is shorter than the `ALT` sequence, it is considered an insertion. For example:

| CHROM | POS  | ID           | REF     | ALT       | FILTER |
| ----- | ---- | ------------ | ------- | --------- | ------ |
| 1     | 2000 | standard_del | CT      | C         | PASS   |
| 1     | 2100 | standard_ins | C       | CAA       | PASS   |
| 1     | 2200 | compund_del  | CCTGAAA | CGA       | PASS   |
| 1     | 2300 | compund_ins  | GT      | CAATATATA | PASS   |

are returned as:

| CHROM | POS  | ID            | REF   | ALT      | FILTER | [`VariantType`](#varianttype) |
| ----- | ---- | ------------- | ----- | -------- | ------ | ----------------------------- |
| 1     | 2000 | standard_del  | CT    | C        | PASS   | DEL                           |
| 1     | 2100 | standard_ins  | C     | CAA      | PASS   | INS                           |
| 1     | 2201 | compund_del_0 | C     | G        | PASS   | SNV                           |
| 1     | 2202 | compund_del_1 | TGAAA | T        | PASS   | DEL                           |
| 1     | 2202 | compund_del_2 | T     | A        | PASS   | SNV                           |
| 1     | 2300 | compund_ins_0 | G     | C        | PASS   | SNV                           |
| 1     | 2301 | compund_ins_1 | T     | TATATATA | PASS   | DEL                           |
| 1     | 2301 | compund_ins_2 | T     | A        | PASS   | SNV                           | --> |
</section>
<section id="structural-variants">
<h3>Structural variants<a class="headerlink" href="#structural-variants" title="Link to this heading"></a></h3>
<p>VariantExtractor returns one entry per structural variant (one entry per breakend pair). This helps to avoid the ambiguity of the notation and keeps the process deterministic. For this reason, in case of paired breakends, the breakend with the lowest chromosome and/or position is returned. If a breakend is not the lowest chromosome and/or position and is missing its pair, its pair is <a class="reference internal" href="#inferred-breakend-pairs"><span class="xref myst">inferred and returned</span></a>.</p>
<section id="breakend-vs-shorthand-notation">
<h4>Breakend vs shorthand notation<a class="headerlink" href="#breakend-vs-shorthand-notation" title="Link to this heading"></a></h4>
<p>Entries with the same information, either described with shorthand or breakend notation, will be returned the same way. Here is an example for a DEL entry:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>3000</p></td>
<td><p>event_1_o</p></td>
<td><p>A</p></td>
<td><p>A[1:5000[</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>5000</p></td>
<td><p>event_1_h</p></td>
<td><p>A</p></td>
<td><p>]1:3000]A</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
<tr class="row-even"><td><p>1</p></td>
<td><p>3000</p></td>
<td><p>event_1</p></td>
<td><p>A</p></td>
<td><p>A[1:5000[</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=DEL</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>3000</p></td>
<td><p>event_1</p></td>
<td><p>A</p></td>
<td><p>&lt;DEL&gt;</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=DEL; END=5000</p></td>
</tr>
<tr class="row-even"><td><p>1</p></td>
<td><p>3000</p></td>
<td><p>event_1</p></td>
<td><p>AGTCACAA…</p></td>
<td><p>A</p></td>
<td><p>PASS</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p>are returned as one entry (each one of them with their own <code class="docutils literal notranslate"><span class="pre">ALT</span></code> field), but with the same <code class="docutils literal notranslate"><span class="pre">VariantRecord.end</span></code> and <code class="docutils literal notranslate"><span class="pre">VariantType</span></code>:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
<th class="head"><p><a class="reference internal" href="#varianttype"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></span></a></p></th>
<th class="head"><p><a class="reference internal" href="#variantrecord"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantRecord.end</span></code></span></a></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>3000</p></td>
<td><p>event_1</p></td>
<td><p>A</p></td>
<td><p>…</p></td>
<td><p>PASS</p></td>
<td><p>…</p></td>
<td><p>DEL</p></td>
<td><p>5000</p></td>
</tr>
</tbody>
</table>
<section id="the-special-case-of-inv">
<h5>The special case of INV<!-- omit in toc --><a class="headerlink" href="#the-special-case-of-inv" title="Link to this heading"></a></h5>
<p>&lt;INV&gt; is a special case of shorthand notation because it represents two paired breakends. For example, the following shorthand notation:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>321682</p></td>
<td><p>event_1</p></td>
<td><p>T</p></td>
<td><p>&lt;INV&gt;</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=INV;END=421681</p></td>
</tr>
</tbody>
</table>
<p>is equivalent to the following breakends:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
<th class="head"><p><a class="reference internal" href="#varianttype"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></span></a></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>321681</p></td>
<td><p>event_1_0</p></td>
<td><p>N</p></td>
<td><p>N]2:421681]</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=INV</p></td>
<td><p>INV</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>321682</p></td>
<td><p>event_1_1</p></td>
<td><p>T</p></td>
<td><p>[2:421682[T</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=INV</p></td>
<td><p>INV</p></td>
</tr>
</tbody>
</table>
<p>In this case, VariantExtractor converts internally &lt;INV&gt; to two entries with breakend notation (one for each breakend pair). Note that the <code class="docutils literal notranslate"><span class="pre">N</span></code> will be replaced with the correct nucleotide if <code class="docutils literal notranslate"><span class="pre">fasta_ref</span></code> is provided to VariantExtractor.</p>
</section>
</section>
<section id="paired-breakends">
<h4>Paired breakends<a class="headerlink" href="#paired-breakends" title="Link to this heading"></a></h4>
<p>For <strong>paired breakends</strong>, breakends are paired using the <code class="docutils literal notranslate"><span class="pre">INFO</span></code> fields <code class="docutils literal notranslate"><span class="pre">MATEID</span></code> or <code class="docutils literal notranslate"><span class="pre">PARID</span></code>. If these fields are not available, they are paired using their coordinates (contig+position). The breakend with the lowest chromosome and/or position is returned. For example:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>3000</p></td>
<td><p>event_1_o</p></td>
<td><p>T</p></td>
<td><p>]3:5000]T</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>5000</p></td>
<td><p>event_1_h</p></td>
<td><p>G</p></td>
<td><p>G[2:3000[</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
<tr class="row-even"><td><p>1</p></td>
<td><p>3000</p></td>
<td><p>event_2_o</p></td>
<td><p>A</p></td>
<td><p>A[1:5000[</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>5000</p></td>
<td><p>event_2_h</p></td>
<td><p>A</p></td>
<td><p>]1:3000]A</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
</tbody>
</table>
<p>are returned as one entry per variant:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
<th class="head"><p><a class="reference internal" href="#varianttype"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></span></a></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>3000</p></td>
<td><p>event_1_o</p></td>
<td><p>T</p></td>
<td><p>]3:5000]T</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
<td><p>TRA</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>3000</p></td>
<td><p>event_2_o</p></td>
<td><p>A</p></td>
<td><p>A[1:5000[</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
<td><p>DEL</p></td>
</tr>
</tbody>
</table>
</section>
<section id="inferred-breakend-pairs">
<h4>Inferred breakend pairs<a class="headerlink" href="#inferred-breakend-pairs" title="Link to this heading"></a></h4>
<p>If <strong>all</strong> breakends are missing their pair, the breakends with the lowest chromosome and/or position are inferred and returned. For example:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>3</p></td>
<td><p>5000</p></td>
<td><p>event_1_h</p></td>
<td><p>G</p></td>
<td><p>G[2:3000[</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>5000</p></td>
<td><p>event_2_h</p></td>
<td><p>A</p></td>
<td><p>]1:3000]A</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
</tbody>
</table>
<p>are returned as their inferred breakend pair with the lowest chromosome and/or position:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
<th class="head"><p><a class="reference internal" href="#varianttype"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></span></a></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>3000</p></td>
<td><p>event_1_h</p></td>
<td><p>N</p></td>
<td><p>]3:5000]N</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
<td><p>TRA</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>3000</p></td>
<td><p>event_2_h</p></td>
<td><p>A</p></td>
<td><p>A[1:5000[</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
<td><p>DEL</p></td>
</tr>
</tbody>
</table>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">N</span></code> will be replaced with the correct nucleotide if <code class="docutils literal notranslate"><span class="pre">fasta_ref</span></code> is provided to VariantExtractor. The following equivalencies are applied:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM1</p></th>
<th class="head"><p>POS1</p></th>
<th class="head"><p>REF1</p></th>
<th class="head"><p>ALT1</p></th>
<th class="head"><p>CHROM2</p></th>
<th class="head"><p>POS2</p></th>
<th class="head"><p>REF2</p></th>
<th class="head"><p>ALT2</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>500</p></td>
<td><p>N</p></td>
<td><p>N[7:800[</p></td>
<td><p>7</p></td>
<td><p>800</p></td>
<td><p>N</p></td>
<td><p>]1:500]N</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>500</p></td>
<td><p>N</p></td>
<td><p>]7:800]N</p></td>
<td><p>7</p></td>
<td><p>800</p></td>
<td><p>N</p></td>
<td><p>N[1:500[</p></td>
</tr>
<tr class="row-even"><td><p>1</p></td>
<td><p>500</p></td>
<td><p>N</p></td>
<td><p>[7:800[N</p></td>
<td><p>7</p></td>
<td><p>800</p></td>
<td><p>N</p></td>
<td><p>[1:500[N</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>500</p></td>
<td><p>N</p></td>
<td><p>N]7:800]</p></td>
<td><p>7</p></td>
<td><p>800</p></td>
<td><p>N</p></td>
<td><p>N]1:500]</p></td>
</tr>
</tbody>
</table>
</section>
<section id="imprecise-paired-breakends">
<h4>Imprecise paired breakends<a class="headerlink" href="#imprecise-paired-breakends" title="Link to this heading"></a></h4>
<p>Imprecise breakends do not match exactly with their pair in coordinates. In this case, they are paired using the <code class="docutils literal notranslate"><span class="pre">INFO</span></code> fields <code class="docutils literal notranslate"><span class="pre">MATEID</span></code> or <code class="docutils literal notranslate"><span class="pre">PARID</span></code>. As with the rest of variants, for each breakend pair, only the breakend with the lowest chromosome and/or position is returned. However, it is important to notice that the <code class="docutils literal notranslate"><span class="pre">CIPOS</span></code> field is lost for the other breakend. For example:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>3010</p></td>
<td><p>event_1_o</p></td>
<td><p>T</p></td>
<td><p>T[3:5000[</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND;CIPOS=0,50;PARID=event_1_h</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>5050</p></td>
<td><p>event_1_h</p></td>
<td><p>A</p></td>
<td><p>]2:3050]A</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND;CIPOS=0,100;PARID=event_1_o</p></td>
</tr>
</tbody>
</table>
<p>are paired and the entry with the lowest chromosome and/or position is returned:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
<th class="head"><p><a class="reference internal" href="#varianttype"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></span></a></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>3010</p></td>
<td><p>event_1_o</p></td>
<td><p>T</p></td>
<td><p>T[3:5000[</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND;CIPOS=0,50;PARID=a_h</p></td>
<td><p>TRA</p></td>
</tr>
</tbody>
</table>
</section>
<section id="single-breakends">
<h4>Single breakends<a class="headerlink" href="#single-breakends" title="Link to this heading"></a></h4>
<p>Single breakends cannot be matched with other breakends because they lack a mate. They may be able to be matched later in downstream analysis. That is why each one is kept as a different variant. For example:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>3000</p></td>
<td><p>event_s</p></td>
<td><p>T</p></td>
<td><p>T.</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>5000</p></td>
<td><p>event_m</p></td>
<td><p>G</p></td>
<td><p>.G</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
</tr>
</tbody>
</table>
<p>are returned as two entries:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CHROM</p></th>
<th class="head"><p>POS</p></th>
<th class="head"><p>ID</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>ALT</p></th>
<th class="head"><p>FILTER</p></th>
<th class="head"><p>INFO</p></th>
<th class="head"><p><a class="reference internal" href="#varianttype"><span class="xref myst"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></span></a></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>2</p></td>
<td><p>3000</p></td>
<td><p>event_s</p></td>
<td><p>T</p></td>
<td><p>T.</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
<td><p>SGL</p></td>
</tr>
<tr class="row-odd"><td><p>3</p></td>
<td><p>5000</p></td>
<td><p>event_m</p></td>
<td><p>G</p></td>
<td><p>.G</p></td>
<td><p>PASS</p></td>
<td><p>SVTYPE=BND</p></td>
<td><p>SGL</p></td>
</tr>
</tbody>
</table>
</section>
</section>
</section>
<section id="dependencies">
<h2>Dependencies<a class="headerlink" href="#dependencies" title="Link to this heading"></a></h2>
<p>The dependencies are covered by their own respective licenses as follows:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/pysam-developers/pysam">Python/Pysam package</a> (MIT license)</p></li>
</ul>
</section>
</section>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="variant_extractor.html">variant_extractor API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="variant_extractor.html#variant_extractor.VariantExtractor"><code class="docutils literal notranslate"><span class="pre">VariantExtractor</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="variant_extractor.html#variant_extractor.variants.BreakendSVRecord"><code class="docutils literal notranslate"><span class="pre">BreakendSVRecord</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="variant_extractor.html#variant_extractor.variants.ShorthandSVRecord"><code class="docutils literal notranslate"><span class="pre">ShorthandSVRecord</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="variant_extractor.html#variant_extractor.variants.VariantRecord"><code class="docutils literal notranslate"><span class="pre">VariantRecord</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="variant_extractor.html#variant_extractor.variants.VariantType"><code class="docutils literal notranslate"><span class="pre">VariantType</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">CHANGELOG</a><ul>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-10-2025-03-26">v4.0.10 (2025-03-26)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-9-2025-03-26">v4.0.9 (2025-03-26)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-8-2024-07-17">v4.0.8 (2024-07-17)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-7-2024-06-28">v4.0.7 (2024-06-28)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-6-2023-03-08">v4.0.6 (2023-03-08)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-5-2023-03-06">v4.0.5 (2023-03-06)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-4-2023-03-02">v4.0.4 (2023-03-02)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-3-2023-03-02">v4.0.3 (2023-03-02)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-2-2023-02-16">v4.0.2 (2023-02-16)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-1-2023-02-15">v4.0.1 (2023-02-15)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v4-0-0-2023-02-10">v4.0.0 (2023-02-10)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v3-1-0-2022-06-14">v3.1.0 (2022-06-14)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v3-0-0-2022-04-25">v3.0.0 (2022-04-25)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v2-1-0-2022-04-20">v2.1.0 (2022-04-20)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#v2-0-0-2022-04-04">v2.0.0 (2022-04-04)</a></li>
</ul>
</li>
</ul>
</div>
<section id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Link to this heading"></a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="variant_extractor.html" class="btn btn-neutral float-right" title="variant_extractor API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Rapsssito.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>